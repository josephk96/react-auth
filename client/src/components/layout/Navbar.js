import React, { useContext } from 'react';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import breakpoint, { map } from "styled-components-breakpoint";

import AuthContext from "../../context/auth/authContext";

const NavContainer = styled.div`
top: 0;
position: sticky;
  width: 100%;
  grid-area: navbar;
`;

const Nav = styled.nav`
  background-color: #60adf0;
  height: 100%;
  display: flex;
  justify-content: center;
  ${breakpoint("md")`
    justify-content: flex-end;
    `}
`;

const LoggedInNav = styled.nav`
  height: 100%;
  background-color: #60adf0;
  display: flex;
  flex-direction: row;
  justify-content: flex-end;
  ${breakpoint("md")`
    justify-content: flex-end;
    `}
`;

const List = styled.ul`
  display: flex;
  align-items: center;
  font-size: 1.5rem;
  justify-content: center;
  font-family: "Lato", sans-serif;
  ${breakpoint("md")`
    justify-content: flex-end;
    margin-top: 1.5rem;
    `}
`;

const StyledLink = styled(Link)`
  text-decoration: none;
  color: white;
  padding: 1.5rem;
  ${breakpoint("md")`
    height: 3.5rem;
    margin-right: 3rem;
    `}
`;

const StyledLogout = styled(Link)`
  text-decoration: none;
  font-size: 3rem;
  color: white;
  padding: 1.5rem;
  justify-self: flex-start;

  ${breakpoint("md")`
    height: 3.5rem;
    margin-right: 3rem;
    `}
`;

StyledLogout.displayName = 'StyledLogout';

const Navbar = () => {
  const authContext = useContext(AuthContext);
  const { isAuthenticated, logout } = authContext;

  return (
    <>
      {isAuthenticated ? (
        <NavContainer>
          <LoggedInNav>
            <List>
              <StyledLogout onClick={() => logout()}>
                <i class="fas fa-sign-out-alt"></i>
              </StyledLogout>
            </List>
          </LoggedInNav>
        </NavContainer>
      ) : (
        <NavContainer>
          <Nav>
            <List>
              <StyledLink to="/login">Login</StyledLink>
              <StyledLink to="/register">Register</StyledLink>
            </List>
          </Nav>
        </NavContainer>
      )}
    </>
  );
}

export default Navbar
