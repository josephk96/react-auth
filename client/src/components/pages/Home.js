import React, {  useState, useContext, useEffect, useLayoutEffect } from "react";
import AuthContext from "../../context/auth/authContext";
import Chat from '../layout/Chat';
import styled from 'styled-components';
import breakpoint, { map } from "styled-components-breakpoint";

import ReconnectingWebSocket from 'reconnecting-websocket';
import axios from 'axios';

const url = "ws://192.168.0.101";
const ws = new ReconnectingWebSocket(url);

// const Container = styled.div`
//   grid-area: body;
//   ${breakpoint("md")`
//   `}
// `;

const ChatContainer = styled.div`
  flex-direction: column;
  justify-content: space-between;
`;

ChatContainer.displayName = 'ChatContainer'

const Form = styled.form`
  grid-area: footer;
  display: block;
  height: 100%;
  bottom: 0;
  position: sticky;
  background-color: #ffffff;
`;

Form.displayName = 'Form';

const InputWrapper = styled.div`
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: row;
`

const Input = styled.input`
  display: flex;
  flex: 9;
  font-size: 20px;
  padding: 2rem;
  border: none;
  outline: none;
  border-radius: 0px;
  ${breakpoint("md")`
    padding: 3rem;
    `}
`;

Input.displayName = 'Input'

const Button = styled.button`
  display: flex;
  padding-top: 0.6rem;
  padding-right: 1rem;
  flex: 1;
  justify-content: center;
  font-size: 2.3rem;
  border: none;
  outline: none;
  background-color: white;
  color: #60adf0;
  ${breakpoint("md")`
    `}
`;

Button.displayName = 'Button';

const Home = () => {
  const authContext = useContext(AuthContext);
  const { verifyLocalStorage, user, loading } = authContext;
  const [message, setMessage] = useState({ msg: '' });
  const [messages, setMessages] = useState([]);

  const { msg } = message;

  const onChange = e => setMessage({ [e.target.name] : e.target.value });

  const addChat = (msg, name, time) => {
    setMessages([...messages, {msg, name, time }]);
  }


  const onSubmit = e => {
    if(msg !== '') {
      e.preventDefault(); 
      let today = new Date();
      let time = today.toLocaleTimeString([], { hour: 'numeric', minute: 'numeric' });
      addChat(msg, localStorage.name, time);
      ws.send(JSON.stringify({msg, name: localStorage.name, time}));
    }
    e.preventDefault();
    setMessage({ msg: '' });
  };

  useEffect(() => {
        ws.onopen = () => console.log(JSON.stringify('WS Connected!'));

        ws.onerror = error => {
          console.log(`WebSocket error:`);
          console.log(error);
        };
                
        ws.onmessage = e => {
          const { msg, name, time } = JSON.parse(e.data);
          if (name !== localStorage.name) {
            addChat(msg, name, time);
          }
        };

        ws.onclose = () => {
          console.log("disconnected!");
        };
  })


  return (
    <>
      {/* <LeftSidebar /> */}
      <ChatContainer>
        <Chat messages={messages} />
      </ChatContainer>
      <Form>
        <InputWrapper>
          <Input
            onChange={onChange}
            value={msg}
            name="msg"
            autoComplete="off"
          />
          <Button type="submit" onClick={onSubmit}>
            <i class="far fa-paper-plane"></i>
          </Button>
        </InputWrapper>
      </Form>
    </>
  );
}

export default Home;
