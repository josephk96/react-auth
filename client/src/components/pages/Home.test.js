import React from "react";
import Home from "./Home";
import Navbar from '../layout/Navbar';
import { shallow } from "enzyme";

jest.mock("react", () => {
  const ActualReact = require.requireActual("react");
  return {
    ...ActualReact,
    useContext: () => ({
      authContext: {}
    })
  };
});

describe("Home", () => {
  const component = shallow(<Home />);

  it("renders without crashing", () => {
    expect(component.debug()).toMatchSnapshot();
  });

  it('has a chat container', () => {
    expect(component.find('ChatContainer')).toHaveLength(1);
  })

  it("has an input field", () => {
    expect(component.find("Input")).toHaveLength(1);
  });

  it("has a send button", () => {
    expect(component.find("Button")).toHaveLength(1);
  });
});
