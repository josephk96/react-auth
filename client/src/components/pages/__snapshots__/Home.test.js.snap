// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders without crashing 1`] = `
"<Fragment>
  <ChatContainer>
    <Chat messages={{...}} />
  </ChatContainer>
  <Form>
    <styled.div>
      <Input onChange={[Function: onChange]} value=\\"\\" name=\\"msg\\" autoComplete=\\"off\\" id=\\"msgInput\\" />
      <Button type=\\"submit\\" onClick={[Function: onSubmit]}>
        <i class=\\"far fa-paper-plane\\" />
      </Button>
    </styled.div>
  </Form>
</Fragment>"
`;
